contracts/oracle/BackedOracle.sol

```diff
@@ -0,0 +1,201 @@
+/**
+ * SPDX-License-Identifier: MIT
+ *
+ * Copyright (c) 2021-2023 Backed Finance AG
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+pragma solidity 0.8.9;
+
+import "./BackedOracleInterface.sol";
+
+import "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+
+contract BackedOracle is AccessControlUpgradeable, AggregatorV2V3Interface {
+    uint8 public constant VERSION = 1;
+    uint8 public constant MAX_PERCENT_DIFFERENCE = 10;
+    uint32 public constant MAX_TIMESTAMP_AGE = 5 minutes;
+    uint32 public constant MIN_UPDATE_INTERVAL = 1 hours;
+
+    bytes32 public constant UPDATER_ROLE = keccak256("UPDATER_ROLE");
+
+    struct RoundData {
+        int192 answer;
+        uint32 timestamp;
+    }
+
+    uint8 private _decimals;
+    string private _description;
+
+    mapping(uint256 => RoundData) private _roundData;
+    uint80 private _latestRoundNumber;
+
+    constructor() {
+        initialize(0, "Backed Oracle Implementation", address(0), address(0));
+    }
+
+    function initialize(
+        uint8 __decimals,
+        string memory __description,
+
+        address __admin,
+        address __updater
+    ) public initializer {
+        _decimals = __decimals;
+        _description = __description;
+
+        _setupRole(DEFAULT_ADMIN_ROLE, __admin);
```

Comment:
_setupRole is actually depreciated in favor of _grantRole. _setupRole just calls _grantRole now but better to use _grantRole then